client/client.go

```diff
@@ -2144,10 +2153,32 @@ func (c *Client) allocSync() {
 			if len(updates) == 0 {
 				continue
 			}
+			// Ensure we never send an update before we've had at least one sync
+			// from the server
+			select {
+			case <-c.serversContactedCh:
+			default:
+				continue
+			}
 
 			sync := make([]*structs.Allocation, 0, len(updates))
-			for _, alloc := range updates {
-				sync = append(sync, alloc)
+			for allocID, update := range updates {
+				if ar, ok := c.allocs[allocID]; ok {
+					if !ar.LastAcknowledgedStateIsCurrent(update) {
+						sync = append(sync, update)
+					}
+				} else {
+					// no allocrunner (typically a failed placement), so we need
+					// to send update
+					sync = append(sync, update)
+				}
+			}
```

Comment:
nit: can you pull this into a helper function like `deduplicateUpdates(updates, allocs) sync`

easier to read and write full coverage tests that way